cmd/vulcan-semgrep/main.go

```diff
@@ -0,0 +1,256 @@
+				return err
+			}
+		}
+
+		logger.WithFields(logrus.Fields{"options": opt}).Debug("using options")
+
+		// We check if the target is not the public Github.
+		targetURL, err := url.Parse(target)
+		if err != nil {
+			return err
+		}
+
+		// TODO: Support multiple authenticated Github Enterprise instances.
+		githubURL, err := url.Parse(os.Getenv("GITHUB_ENTERPRISE_ENDPOINT"))
+		if err != nil {
+			return err
+		}
+
+		var auth *http.BasicAuth
+		if githubURL.Host != "" && targetURL.Host == githubURL.Host {
+			auth = &http.BasicAuth{
+				Username: "username", // Can be anything except blank.
+				Password: os.Getenv("GITHUB_ENTERPRISE_TOKEN"),
+			}
+			logger.Debug("using credentials for GitHub")
+		}
+
+		gitCreds := &helpers.GitCreds{}
+		if auth != nil {
+			gitCreds.User = auth.Username
+			gitCreds.Pass = auth.Password
+		}
+		isReachable, err := helpers.IsReachable(target, assetType, gitCreds)
+		if err != nil {
+			logger.WithError(err).Warn("can not check asset reachability")
+		}
+		if !isReachable {
+			return checkstate.ErrAssetUnreachable
+		}
+
+		repoPath := filepath.Join("/tmp", filepath.Base(targetURL.Path))
+		if err := os.Mkdir(repoPath, 0755); err != nil {
+			return err
+		}
+
+		logger.WithFields(logrus.Fields{"repo_path": repoPath}).Debug("cloning repo")
+
+		_, err = git.PlainClone(repoPath, false, &git.CloneOptions{
+			URL:   target,
+			Auth:  auth,
+			Depth: opt.Depth,
+		})
+		if err != nil {
+			return err
+		}
+
+		r, err := runSemgrep(ctx, logger, opt.Ruleset, repoPath)
+		if err != nil {
+			return err
+		}
+
+		addVulnsToState(state, r, repoPath, target)
+
+		return nil
+	}
+
+	c := check.NewCheckFromHandler(checkName, run)
+	c.RunAndServe()
+}
+
+func addVulnsToState(state checkstate.State, r *SemgrepOutput, repoPath, target string) {
+	if r == nil || len(r.Results) < 1 {
+		return
+	}
+
+	logger.WithFields(logrus.Fields{"num_results": len(r.Results)}).Info("processing results")
+
+	vulns := make(map[string]report.Vulnerability)
+	for _, result := range r.Results {
+		filepath := strings.TrimPrefix(result.Path, fmt.Sprintf("%s/", repoPath))
+		path := fmt.Sprintf("%s:%d", filepath, result.Start.Line)
+
+		v := vuln(result, filepath, vulns)
+
+		row := map[string]string{
+			"Path":  path,
+			"Match": result.Extra.Lines,
+			"Fix":   result.Extra.Fix,
+			// Message will be removed afterwards if it has the same value in
+			// all the rows.
+			"Message": result.Extra.Message,
+		}
+
+		// In almost all cases the message is the same for all the results of
+		// the same rule, but there are few cases where message differs. In
+		// those few cases we will be adding the alternative messages in the
+		// resources table.
+		if result.Extra.Message != v.Description && len(v.Resources[0].Header) == 3 {
+			logger.WithFields(logrus.Fields{"vulnerability": v.Summary}).Info("vulnerability has alternative messages")
+
+			v.Resources[0].Header = append(v.Resources[0].Header, "Message")
+			v.Description = ""
```

Comment:
If I understood correctly, "initially" we define the vulnerability `Description` as the `result.Extra.Message`. But then, if one occurrence of the same vulnerability in the same file has a different message, then we set the `Description` as `""` and collect these different messages in the resources table. Is this correct?

Regarding this, have you thought about how can this impact the opening/fixing criteria at the Vulnerability DB level which is based on "summary+description" to compare issues?

Could this lead to potentially undesired behavior? For example an issue that on the first iteration is defined with a `Description` field but on a posterior iteration, maybe because in this case it is found in more than one place and for some reason the `Extra.Message` is different for the occurrences, the description is set to void. In this case this will fix the finding created in the first iteration and create a new one I believe.
Maybe it's best to just not define a description at all? I don't know how much probable is this situation to happen, you probably have a better context than me in this regard.

Reply:
Agree, I think that's our best option considering the possible variations in the description field and its implications at the VulnDB level.